### PR TITLE
Implemented TraditionalModel using XGBoost with model persistence and evaluation

### LINKED ISSUE #2: Traditional approach
Traditional ML approach for recommending recipes

### PR DESCRIPTION
**Pull request by johnjoel2001**
**PR Description:**

This pull request introduces the `TraditionalModel`, a collaborative filtering approach leveraging` XGBoost` for rating prediction. Key enhancements include model training, evaluation, and storage for reuse.

**What’s Added:**

- `TraditionalModel` implemented using `XGBRegressor` for improved accuracy on the user–recipe rating prediction.

- Achieved RMSE: 1.3763 on the test set.

- Introduced `LabelEncoder` usage for encoding `user_id` and `recipe_id`.

- **Model persistence:** Trained models and encoders are now saved to a new directory: `scripts/models/.`

- Automatic fallback logic in `get_instance()` to load the model if available, or train and save a new one if not.

**New Files/Directories:**
- `scripts/models`/: Stores serialized model and encoders:
    
    - `user_encoder.pkl`
    - `item_encoder.pkl`
    - `traditional_model.pkl`

**How It Works:**
- The model uses user-item pairs as input features, encodes them, and trains a regression model using `XGBoost`.

- At runtime, the model is loaded if already trained. Otherwise, it trains and saves automatically.

- Evaluation is supported via the `evaluate()` method and integrated with `eval.py` for performance comparison.

**Note:**

After merging the previous branch into `main`, it was observed that updates to `etl.py`, `naive.py`, and `eval.py` did not reflect in the current `traditional` branch. To maintain consistency and ensure full functionality, those changes were manually updated in this branch.

### CODE REVIEW

File: scripts/eval/evaluate.py [hunk 1]
```diff
@@ -1,10 +1,16 @@
 import pandas as pd
-from ..etl.etl import extract
-from ..util.model import Model
-from ..etl.etl import train_test_split_reviews
-from ..naive.model import NaiveModel
-from ..traditional.model import TraditionalModel
-from ..deep.model import DeepLearningModel
```
Comment (haranku16) [inline scripts/eval/evaluate.py]: I actually find this syntax to be more straightforward. If you run the script as described in the readme (as a module)

```
python -m scripts.eval.evaluate
```

Then these relative imports work without needing to append to the system path

**Review by haranku16 — CHANGES_REQUESTED** [review]
Thanks for working on this John, though I'm concerned that the XGBoost algorithm isn't really learning anything if we only supply recipe and user ids without additional features. This could be why the traditional approach is performing worse than the naive approach. Perhaps if we used a hybrid KNN-XGBoost approach, we could get better performance out of this model, or use another traditional approach like SVD for collaborative filtering. Here's an [example notebook](https://github.com/zas03ia/Recommender-System-with-Collaborative-Filtering-and-XGBoost/blob/main/Recommender%20System.ipynb) that could come in handy.

**Review by johnjoel2001 — COMMENTED** [review]
Thank you @haranku16 for the helpful feedback - it was spot on and led to significant improvements.

After reviewing the initial **XGBoost** based model, I realized that using only `user_id `and `recipe_id `as features severely limited the model’s ability to generalize. As you suggested, I explored alternative approaches to improve performance, including **KNN-based regression, SVD, and hybrid options.**

The most promising result came from a feature-enriched KNN regression model that outperformed the naive baseline.

**What’s improved:**

**Feature Engineering:**  Added meaningful features like each user's average rating (`user_avg_rating`) and each recipe’s average rating (`recipe_avg_rating`) alongside encoded IDs.

**Preprocessing:** Standardized all features using `StandardScaler `to optimize performance for distance-based models.

**Hyperparameter Tuning:** Used `GridSearchCV `to search across neighbors, distance metrics, and weight types to find the best configuration.

**Outcome**: The updated model achieved an **RMSE of 0.5827**, a significant improvement over both the naive model and the original XGBoost implementation. 

Screenshot 2025-04-14 081932

Thanks again - your suggestion was a key turning point in making the model more robust and predictive.

File: scripts/eval/evaluate.py (hunk 1 above)
Comment (johnjoel2001) [reply scripts/eval/evaluate.py]: Yes @haranku16, I agree. I have removed the system path in all the scripts.

**Review by haranku16 — APPROVED** [review]
Thanks for the modified approach which surpasses the naive approach in its performance! I see that there are a few merge conflicts that need to be resolved. Let's please merge the latest from main into your branch and push another commit so this PR can be merged.

```
git fetch
git merge origin/main
```

**Review by haranku16 — APPROVED** [review]
